### PR TITLE
[HOPS-1541] provenance

### PR DESCRIPTION
**Pull request by o-alex**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added and passed (for bug fixes / features)
- [X] HOPS JIRA issue has been opened for this PR
- [X] All commits have been squashed down to a single commit
- [X] The commit message has the following format: [HOPS-XXX] message

* **Post a link to the associated JIRA issue**
https://hopshadoop.atlassian.net/browse/HOPS-1541

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature - provenance

* **What is the new behavior (if this is a feature change)?**
Log filesystem operations into a log tailed by epipe

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
This JIRA has changes to the database (add tables)

* **Other information**:

### CODE REVIEW

File: src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvXAttrBufferClusterj.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  private FileProvXAttrBufferEntry createProvenanceLogEntry(FileProvXAttrBufferEntryDto dto) {
+    return new FileProvXAttrBufferEntry(
+      dto.getINodeId(),
+      dto.getNamespace(),
+      dto.getName(),
+      dto.getINodeLogicalTime(),
+      dto.getValue());
+  }
```
Comment (berthoug) [inline src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvXAttrBufferClusterj.java]: Is there a plane to use these functions in the future? Write now they are not usefull as we only write to the database and never read from it.

File: src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvenanceClusterj.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+      dto.getDatasetName(),
+      dto.getP1Name(),
+      dto.getP2Name(),
+      dto.getParentName(),
+      dto.getUserName(),
+      dto.getXAttrName(),
+      dto.getLogicalTimeBatch(),
+      dto.getLogicalTimeBatch(),
+      dto.getDsLogicalTime());
+  }
```
Comment (berthoug) [inline src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvenanceClusterj.java]: Is there a plane to use these functions in the future? Write now they are not usefull as we only write to the database and never read from it.

File: src/main/java/io/hops/metadata/ndb/dalimpl/yarn/AppProvenanceClusterJ.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    return new AppProvenanceEntry(
+      dto.getId(),
+      dto.getName(),
+      dto.getState(),
+      dto.getUser(),
+      dto.getTimestamp(),
+      dto.getSubmitTime(), 
+      dto.getStartTime(),
+      dto.getFinishTime());
+  }
```
Comment (berthoug) [inline src/main/java/io/hops/metadata/ndb/dalimpl/yarn/AppProvenanceClusterJ.java]: Is there a plane to use these functions in the future? Write now they are not usefull as we only write to the database and never read from it.

File: src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvXAttrBufferClusterj.java (hunk 1 above)
Comment (o-alex) [reply src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvXAttrBufferClusterj.java]: Fixed, removed createCollection and createProvenanceLogEntry - it was dead code

File: src/main/java/io/hops/metadata/ndb/ClusterjConnector.java [hunk 4]
```diff
@@ -34,6 +34,7 @@
 import io.hops.metadata.ndb.mysqlserver.MysqlServerConnector;
 import io.hops.metadata.ndb.wrapper.HopsSession;
 import io.hops.metadata.ndb.wrapper.HopsTransaction;
+import io.hops.metadata.yarn.dal.AppProvenanceDataAccess;
```
Comment (maismail) [inline src/main/java/io/hops/metadata/ndb/ClusterjConnector.java]: It seems you missed to import the other two classes 
FileProvenanceDataAccess  and FileProvXAttrBufferDataAccess

File: src/main/java/io/hops/metadata/ndb/NdbStorageFactory.java [hunk 5]
```diff
@@ -52,6 +53,7 @@
 import io.hops.metadata.ndb.dalimpl.yarn.rmstatestore.DelegationKeyClusterJ;
 import io.hops.metadata.ndb.dalimpl.yarn.rmstatestore.DelegationTokenClusterJ;
 import io.hops.metadata.ndb.mysqlserver.MysqlServerConnector;
+import io.hops.metadata.yarn.dal.AppProvenanceDataAccess;
```
Comment (maismail) [inline src/main/java/io/hops/metadata/ndb/NdbStorageFactory.java]: It seems you missed to import the other two classes
FileProvenanceDataAccess and FileProvXAttrBufferDataAccess and the their clusterj classes

File: src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvenanceClusterj.java (hunk 2 above)
Comment (o-alex) [reply src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvenanceClusterj.java]: fixed - removed the methods

File: src/main/java/io/hops/metadata/ndb/dalimpl/yarn/AppProvenanceClusterJ.java (hunk 3 above)
Comment (o-alex) [reply src/main/java/io/hops/metadata/ndb/dalimpl/yarn/AppProvenanceClusterJ.java]: fixed

File: src/main/java/io/hops/metadata/ndb/ClusterjConnector.java [hunk 6]
```diff
@@ -364,7 +365,8 @@ private boolean formatHDFS(boolean transactional) throws StorageException{
         HashBucketDataAccess.class, StorageDataAccess.class,
         AceDataAccess.class, RetryCacheEntryDataAccess.class, CacheDirectiveDataAccess.class,
         CachePoolDataAccess.class, CachedBlockDataAccess.class,
-        ActiveBlockReportsDataAccess.class, XAttrDataAccess.class, EncryptionZoneDataAccess.class);
+        ActiveBlockReportsDataAccess.class, XAttrDataAccess.class, EncryptionZoneDataAccess.class,
+        FileProvenanceDataAccess.class, FileProvXAttrBufferDataAccess.class, AppProvenanceDataAccess.class);
```
Comment (berthoug) [inline src/main/java/io/hops/metadata/ndb/ClusterjConnector.java]: can you move the AppProvenanceDataAccess to the formatYarn function

File: src/main/java/io/hops/metadata/ndb/ClusterjConnector.java (hunk 4 above)
Comment (o-alex) [reply src/main/java/io/hops/metadata/ndb/ClusterjConnector.java]: there is an import for all the hdfs DALs

https://github.com/hopshadoop/hops-metadata-dal-impl-ndb/blob/acada8c148133fff5c8df49a5c341aadd8d4e86e/src/main/java/io/hops/metadata/ndb/ClusterjConnector.java#L30

File: src/main/java/io/hops/metadata/ndb/NdbStorageFactory.java (hunk 5 above)
Comment (berthoug) [reply src/main/java/io/hops/metadata/ndb/NdbStorageFactory.java]: They are imported by 
import io.hops.metadata.hdfs.dal.*;

File: src/main/java/io/hops/metadata/ndb/ClusterjConnector.java (hunk 4 above)
Comment (berthoug) [reply src/main/java/io/hops/metadata/ndb/ClusterjConnector.java]: import io.hops.metadata.hdfs.dal.*;

File: src/main/java/io/hops/metadata/ndb/NdbStorageFactory.java (hunk 5 above)
Comment (o-alex) [reply src/main/java/io/hops/metadata/ndb/NdbStorageFactory.java]: there is an import for all the hdfs DALs

https://github.com/hopshadoop/hops-metadata-dal-impl-ndb/blob/acada8c148133fff5c8df49a5c341aadd8d4e86e/src/main/java/io/hops/metadata/ndb/NdbStorageFactory.java#L28

File: src/main/java/io/hops/metadata/ndb/ClusterjConnector.java (hunk 6 above)
Comment (o-alex) [reply src/main/java/io/hops/metadata/ndb/ClusterjConnector.java]: fixed

File: src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvXAttrBufferClusterj.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    void setName(String name);
+    
+    @PrimaryKey
+    @Column(name = INODE_LOGICAL_TIME)
+    int getINodeLogicalTime();
+
+    void setINodeLogicalTime(int inodeLogicalTime);
+
+    @Column(name = VALUE)
+    String getValue();
```
Comment (maismail) [inline src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvXAttrBufferClusterj.java]: You should use byte[] for the xattr value. Check the XAttrClusterJ

Comment (o-alex) [reply src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvXAttrBufferClusterj.java]: fixed

File: src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvenanceClusterj.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    @Column(name = USER_ID)
+    int getUserId();
+
+    void setUserId(int userId);
+  
+    //This field is used in case two namenodes/threads, happen to perform a read operation on the same inode,
+    //at the same time(logical and timestamp) on behalf of same user(and app).
+    //We want to be able to store both occurrences,
+    //since the dataset inode logical time might change(change of provenance xattr)
+    //This is an auto increment field used by the processor of the file prov logs(ePipe)
```
Comment (berthoug) [inline src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvenanceClusterj.java]: This should be updated.

Comment (o-alex) [reply src/main/java/io/hops/metadata/ndb/dalimpl/hdfs/FileProvenanceClusterj.java]: modified. do you think it is clear enough now?

**Pull request merged**